torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -762,44 +794,155 @@ def _device_to_params_per_rank(
-            "information has been collected"
+        assert overlap_info.status == _OverlapStatus.INITIALIZED
 
+        self._bucket_assignments_per_rank_cache = [{} for _ in range(self.world_size)]
         params_per_bucket = overlap_info.params_per_bucket
-        for bucket_idx, bucket_params in enumerate(params_per_bucket):
-            assert len(bucket_params) > 0, "Empty bucket"
-            rank = self._ddp_bucket_index_to_rank(bucket_idx)
-            bucket = _DDPBucket(bucket_idx, bucket_params)
-            device = bucket_params[0].device  # assume same device per bucket
-            if device not in self._device_to_buckets_cache:
-                self._device_to_buckets_cache[device] = [[] for _ in range(self.world_size)]
-            self._device_to_buckets_cache[device][rank].append(bucket)
 
-        return self._device_to_buckets_cache
+        if overlap_info.uniform_bucket_assignment:
+            # Define the assignment threshold to approximate uniformity
+            assert overlap_info.total_size is not None, \
+                "`total_size` was not computed"
+            threshold = overlap_info.total_size / self.world_size  # type: ignore[operator]
+            size_per_rank = [0 for _ in range(self.world_size)]
+
+        num_buckets = len(params_per_bucket)
+        overlap_info.assigned_ranks_per_bucket = [set() for _ in range(num_buckets)]
+        assigned_ranks_per_bucket = overlap_info.assigned_ranks_per_bucket
+        if not overlap_info.uniform_bucket_assignment:
+            for bucket_index, bucket_params in enumerate(params_per_bucket):
+                # Assign the entire DDP bucket to a single rank
+                assert len(bucket_params) > 0, "Empty bucket"
+                assigned_rank = self._get_assigned_rank(bucket_index)
+                self._assign_bucket_subset_to_rank(
+                    bucket_index,
+                    bucket_params,
+                    0,
+                    assigned_rank,
+                    assigned_ranks_per_bucket,
+                )
+        else:
+            params_per_bucket_enum = sorted(
+                list(enumerate(params_per_bucket)),
+                key=lambda x: sum([p.numel() for p in x[1]])
+            )
+            for bucket_index, bucket_params in params_per_bucket_enum:
+                assert len(bucket_params) > 0, "Empty bucket"
+                # Assign the DDP bucket to possibly multiple ranks, determined
+                # greedily and subject to the capacity given by `threshold`
+                bucket_offset = 0
+                assignment_size = 0
+                for param_index, param in enumerate(bucket_params):
+                    param_numel = param.numel()
+                    if assignment_size + param_numel >= threshold and param_index > bucket_offset:
+                        assigned_rank = self._get_min_index(size_per_rank, assigned_ranks_per_bucket[bucket_index])
```

Comment:
> What prevents the same rank been returned for different buckets?

Correct, a single rank can now have multiple ZeRO buckets. 

> none of those ZeRO buckets can launch the broadcast before the last DDP bucket is ready.

I do not quite follow why this is. Perhaps I need to think about it more.

The "unit" of each broadcast is a `_DDPBucketAssignment.tensor`, which is the flattened data of that `_DDPBucketAssignment`. Notably, one `_DDPBucketAssignment` does not include parameters across multiple DDP buckets, but a rank may have multiple `_DDPBucketAssignment`s.
- In the non-interleaved approach, we iterate over all bucket indices in increasing order (`0`, ..., `num_buckets - 1`) once the backward computation has finished. For each bucket index, we check if the calling rank has a corresponding `_DDPBucketAssignment` (there may be multiple such ranks). If the calling rank does have one, then it calls `wait()` on that DDP bucket's all-reduce future, performs the partial optimizer step, and launches the broadcast for the `_DDPBucketAssignment`. If there are multiple ranks assigned to the DDP bucket, then we make sure to launch the broadcasts in the same order on all ranks.
This generalizes the existing approach, only that there may be multiple broadcasts for each DDP bucket. In that case, I thought that now if **all** source ranks call `wait()` (compared to before, **one** source rank calls `wait()`), then that would be sufficient. I do not see why there may be a synchronization dependency on a later DDP bucket.
- For the interleaved approach, the pattern of generalization is the same. The difference is there should be no need for an explicit call to `wait()` since the `then()` callback should handle it for us. Again, for each bucket index, we check if the calling rank has a corresponding `_DDPBucketAssignment`, and if so, then it performs the partial optimizer step and launches the broadcast.

Just a guess, is it possible that you think there is only one ZeRO bucket per rank? In that case, I agree that permuting the parameters could cause a rank to have parameters (for example) in the first and last bucket, meaning the rank would have to wait until the last bucket before launching its broadcast. To be clear, now a rank can have multiple ZeRO buckets. This is partially why I mentioned a second constraint to keep in mind is how to keep the number of ZeRO buckets per rank low to avoid too many broadcasts.

The reason I guessed the above is that you particularly singled out the last DDP bucket in your example. However, your argument holds for all DDP buckets, not just the last one. Previously, there was a single rank assigned to update the parameters of a DDP bucket, so only it had to wait for that DDP bucket to be ready. Now, there can be multiple assigned, so all of them must wait for that DDP bucket to be ready. This is the tradeoff for having a potentially more even partitioning. Ultimately, each broadcast for a set of parameters should only need to wait for the one DDP bucket corresponding to those parameters to be ready.

---

Your proposed algorithm seems to constrain ourselves to a single ZeRO bucket per rank (i.e. each rank only serves as the broadcast source once). I am assuming that each ZeRO bucket includes the overflowing parameter. I would be curious how it performs in comparison.

Consider a rank `r` that is assigned parameters `p_1`, ..., `p_k`. Let `I_r` be the set `⋃_{j=1}^{k}` `{bucket_index(p_j)}`, where `bucket_index(p_j)` gives the index of the DDP bucket corresponding to `p_j`. In other words, `I_r` includes exactly the bucket indices of the buckets for which rank `r` has parameters. 
My implementation launches a separate broadcast for every `i` in `I_r`, each right after bucket `i` is ready, while yours only launches one broadcast, when bucket `max(I_r)` is ready. If the set `{p_j}_{j=1}^{k}` is (approximately) the same in both algorithms, then mine could finish earlier since the last broadcast (for bucket `max(I_r)`) communicates possibly fewer parameters than the single broadcast in yours. However, this is relying on the strong assumption that both achieve approximately the same-sized partitioning, which is not always true. In your algorithm, a partition can exceed the uniform partition by at most the size of the largest parameter, while in mine, there is no such bound -- instead, it depends on the number of DDP buckets.

I agree that your algorithm is simpler and may yield a more uniform partitioning. Those two factors make it quite favorable, especially the latter since it reflects the spirit of `ZeroRedundancyOptimizer`. 
IIUC, `uniform_bucket_assignment=False` (existing impl.) < `uniform_bucket_assignment=True` (my new impl.) < your algorithm in terms of uniformity of the partitioning. As such, if we were to only provide a `bool` flag (i.e. two options) to the user, then it may make more sense to have the two extreme points. Although, I do wonder if keeping the existing `uniform_bucket_assignment=False` even makes sense. Indeed, it can achieve better performance in high bandwidth cases since there may be fewer broadcasts, but it certainly violates the spirit of `ZeroRedundancyOptimizer` since it permits highly skewed partitions.

Reply:
> Just a guess, is it possible that you think there is only one ZeRO bucket per rank? ... To be clear, now a rank can have multiple ZeRO buckets. 

I see. Yep, I was assume there is only one ZeRO broadcast per rank, so that the permutation will have an impact on when the ZeRO broadcast can be launched. Would I be correct if I assume that, one rank can launch its multiple ZeRO buckets independently. More specifically, if one rank is responsible for some portion of two DDP buckets x and y, when x is ready, that rank and immediately launch the ZeRO broadcast for its portion in x and does not need to wait for y at all?

Reply:
Yup, exactly.

Reply:
>  Lastly, I am curious if you think there is reason to keep the existing implementation that restricts each ZeRO bucket to exactly match one DDP bucket, if we implement the new algorithm.

I'd suggest let's keep it for now, as we don't know what perf the two algorithms can offer yet, and we could have a private flag to toggle it or making `uniform_bucket_assignment` an Enum (if that's doable). We have already hit the situation previously when working on the two hook algorithm, where the perf disagrees with our hypothesis. It could happen again here as well. We can run some benchmark on popular models to help us make the decision.